File: app/src/main/java/com/example/android/androidskeletonapp/ui/login/LoginActivity.java
package com.example.android.androidskeletonapp.ui.login;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.View;
import android.view.inputmethod.EditorInfo;
import android.widget.ProgressBar;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.ViewModelProvider;

import com.example.android.androidskeletonapp.R;
import com.example.android.androidskeletonapp.data.Sdk;
import com.example.android.androidskeletonapp.data.service.ActivityStarter;
import com.example.android.androidskeletonapp.ui.main.MainActivity;
import com.example.android.androidskeletonapp.ui.programs.ProgramsActivity;
import com.google.android.material.button.MaterialButton;
import com.google.android.material.textfield.TextInputEditText;

import io.reactivex.disposables.Disposable;

public class LoginActivity extends AppCompatActivity {

    private LoginViewModel loginViewModel;
    private Disposable disposable;

    private TextInputEditText serverUrlEditText;
    private TextInputEditText usernameEditText;
    private TextInputEditText passwordEditText;
    private MaterialButton loginButton;
    private ProgressBar loadingProgressBar;

    public static Intent getLoginActivityIntent(Context context) {
        return new Intent(context,LoginActivity.class);
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_login);
        loginViewModel = new ViewModelProvider(this, new LoginViewModelFactory()).get(LoginViewModel.class);

        serverUrlEditText = findViewById(R.id.urlText);
        usernameEditText = findViewById(R.id.usernameText);
        passwordEditText = findViewById(R.id.passwordText);
        loginButton = findViewById(R.id.loginButton);
        loadingProgressBar = findViewById(R.id.loginProgressBar);

        loginViewModel.getLoginFormState().observe(this, loginFormState -> {
            if (loginFormState == null) {
                return;
            }
            loginButton.setEnabled(loginFormState.isDataValid());
            if (loginFormState.getServerUrlError() != null) {
                serverUrlEditText.setError(getString(loginFormState.getServerUrlError()));
            }
            if (loginFormState.getUsernameError() != null) {
                usernameEditText.setError(getString(loginFormState.getUsernameError()));
            }
            if (loginFormState.getPasswordError() != null) {
                passwordEditText.setError(getString(loginFormState.getPasswordError()));
            }
        });

        loginViewModel.getLoginResult().observe(this, loginResult -> {
            if (loginResult == null) {
                return;
            }
            loadingProgressBar.setVisibility(View.GONE);
            if (loginResult.getError() != null) {
                showLoginFailed(loginResult.getError());
            }
            if (loginResult.getSuccess() != null) {
                if (Sdk.d2().programModule().programs().blockingCount() > 0) {
                    ActivityStarter.startActivity(this, ProgramsActivity.getProgramActivityIntent(this),true);
                } else {
                    ActivityStarter.startActivity(this, MainActivity.getMainActivityIntent(this),true);
                }
            }
            setResult(Activity.RESULT_OK);
        });

        TextWatcher afterTextChangedListener = new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
                // ignore
            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                // ignore
            }

            @Override
            public void afterTextChanged(Editable s) {
                loginViewModel.loginDataChanged(
                        serverUrlEditText.getText().toString(),
                        usernameEditText.getText().toString(),
                        passwordEditText.getText().toString());
            }
        };
        serverUrlEditText.addTextChangedListener(afterTextChangedListener);
        usernameEditText.addTextChangedListener(afterTextChangedListener);
        passwordEditText.addTextChangedListener(afterTextChangedListener);

        passwordEditText.setOnEditorActionListener((v, actionId, event) -> {
            if (actionId == EditorInfo.IME_ACTION_DONE) {
                login();
            }
            return false;
        });

        loginButton.setOnClickListener(v -> login());
    }

    private void login() {
        loadingProgressBar.setVisibility(View.VISIBLE);
        loginButton.setVisibility(View.INVISIBLE);
        String username = usernameEditText.getText().toString();
        String password = passwordEditText.getText().toString();
        String serverUrl = serverUrlEditText.getText().toString();

        disposable = loginViewModel
                .login(username, password, serverUrl)
                .doOnTerminate(() -> loginButton.setVisibility(View.VISIBLE))
                .subscribe(u -> {}, t -> {});
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        if (disposable != null) {
            disposable.dispose();
        }
    }

    private void showLoginFailed(String errorString) {
        Toast.makeText(getApplicationContext(), errorString, Toast.LENGTH_SHORT).show();
    }
}

File: app/src/main/java/com/example/android/androidskeletonapp/ui/login/LoginActivity.kt
package com.example.android.androidskeletonapp.ui.login

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.material.Icon
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.outlined.Login
import androidx.compose.runtime.MutableState
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.text.TextRange
import androidx.compose.ui.text.input.TextFieldValue
import androidx.compose.ui.unit.dp
import androidx.databinding.DataBindingUtil
import androidx.lifecycle.ViewModelProvider
import com.example.android.androidskeletonapp.R
import com.example.android.androidskeletonapp.data.Sdk
import com.example.android.androidskeletonapp.data.service.ActivityStarter
import com.example.android.androidskeletonapp.databinding.ActivityLoginBinding
import com.example.android.androidskeletonapp.ui.main.MainActivity
import com.example.android.androidskeletonapp.ui.programs.ProgramsActivity
import io.reactivex.disposables.CompositeDisposable
import org.hisp.dhis.mobile.ui.designsystem.component.Button
import org.hisp.dhis.mobile.ui.designsystem.component.ButtonStyle
import org.hisp.dhis.mobile.ui.designsystem.component.InputShellState
import org.hisp.dhis.mobile.ui.designsystem.component.ProgressIndicator
import org.hisp.dhis.mobile.ui.designsystem.component.ProgressIndicatorType
import org.hisp.dhis.mobile.ui.designsystem.theme.Spacing
import org.hisp.dhis.mobile.ui.designsystem.theme.TextColor

class LoginActivity : AppCompatActivity() {
    private lateinit var binding: ActivityLoginBinding

    private var loginViewModel: LoginViewModel? = null
    private var disposable = CompositeDisposable()
    private var isLoading: MutableState<Boolean> = mutableStateOf(false)
    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        binding = DataBindingUtil.setContentView(this, R.layout.activity_login)

        loginViewModel = ViewModelProvider(this, LoginViewModelFactory()).get(
            LoginViewModel::class.java,
        )

        binding.composeView.setContent {
            val loginUiState by loginViewModel!!.loginUiState.collectAsState()
            var serverUrlText by remember { mutableStateOf(TextFieldValue(getString(R.string.auto_fill_url), TextRange(getString(R.string.auto_fill_url).length))) }
            var userName by remember { mutableStateOf(TextFieldValue(getString(R.string.auto_fill_username), TextRange(getString(R.string.auto_fill_username).length))) }
            var password by remember { mutableStateOf(TextFieldValue(getString(R.string.auto_fill_password), TextRange(getString(R.string.auto_fill_password).length))) }
            var showProgress by remember(isLoading) { mutableStateOf(isLoading) }
            loginViewModel!!.initLoginDefaultValues(
                serverUrlText.text,
                userName.text,
                password.text,
            )
            val isLoginEnabled by remember(loginUiState) { mutableStateOf(loginUiState.isLoginEnabled()) }

            LoginScreen(
                title = getString(R.string.app_name),
                modifier = Modifier,
                icon = {
                    Icon(
                        painter = painterResource(id = R.drawable.ic_dhis2_icon_white),
                        contentDescription = "DHIS2 icon",
                        tint = TextColor.OnPrimary,
                    )
                },
                content = {
                    Box(modifier = Modifier.fillMaxSize()) {
                        Column(
                            modifier = Modifier
                                .fillMaxSize(1f)
                                .padding(horizontal = Spacing.Spacing24),
                            verticalArrangement = Arrangement.Top,
                            horizontalAlignment = Alignment.CenterHorizontally,
                        ) {
                            ProvideInputQRCode(
                                title = getString(R.string.prompt_server_url),
                                state = if (!loginUiState.isServerUrlValid()) InputShellState.ERROR else InputShellState.UNFOCUSED,
                                inputTextFieldValue = serverUrlText,
                                onValueChanged = {
                                    serverUrlText = it
                                    loginViewModel!!.setServer(it.text)
                                },
                                onQRButtonClicked = {},
                                autoCompleteList = getServerAutoCompleteList(),
                            )

                            ProvideInputUser(
                                title = getString(R.string.prompt_username),
                                state = if (!loginUiState.isUserNameValid()) InputShellState.ERROR else InputShellState.UNFOCUSED,
                                inputTextFieldValue = userName,
                                onValueChanged =
                                {
                                    userName = it
                                    loginViewModel!!.setUserName(it.text)
                                },

                            )

                            ProvideInputPassword(
                                title = getString(R.string.prompt_password),
                                state = if (!loginUiState.isPasswordValid()) InputShellState.ERROR else InputShellState.UNFOCUSED,
                                inputTextFieldValue = password,
                                onValueChanged = {
                                    password = it
                                    loginViewModel!!.setPassword(it.text)
                                },

                            )

                            Button(
                                style = ButtonStyle.FILLED,
                                onClick = {
                                    isLoading.value = true
                                    loginViewModel!!.login(loginUiState)
                                },
                                text = getString(R.string.action_sign_in_short),
                                icon = {
                                    Icon(
                                        imageVector = Icons.Outlined.Login,
                                        contentDescription = "Login button",
                                        tint = TextColor.OnPrimary,
                                    )
                                },
                                modifier = Modifier.fillMaxWidth(),
                                enabled = isLoginEnabled,
                            )
                        }
                        if (showProgress.value) {
                            Box(
                                modifier = Modifier
                                    .fillMaxSize(1f)
                                    .background(color = Color.Black.copy(0.1f)),
                            ) {
                                ProgressIndicator(
                                    type = ProgressIndicatorType.CIRCULAR,
                                    modifier = Modifier
                                        .size(80.dp)
                                        .align(Alignment.Center),
                                )
                            }
                        }
                    }
                },
            )
        }
        observeLoginResult()
    }

    private fun observeLoginResult() {
        loginViewModel!!.loginResult.observe(this) { loginResult: LoginResult? ->
            if (loginResult == null) {
                return@observe
            }
            isLoading.value = false
            if (loginResult.error != null) {
                showLoginFailed(loginResult.error)
            }
            if (loginResult.success != null) {
                if (Sdk.d2().programModule().programs().blockingCount() > 0) {
                    ActivityStarter.startActivity(
                        this,
                        ProgramsActivity.getProgramActivityIntent(this),
                        true,
                    )
                } else {
                    ActivityStarter.startActivity(
                        this,
                        MainActivity.getMainActivityIntent(this),
                        true,
                    )
                }
            }
            setResult(RESULT_OK)
        }
    }

    private fun getServerAutoCompleteList(): List<String> {
        return listOf(
            getString(R.string.auto_fill_url),
            getString(R.string.auto_fill_url2),
            getString(
                R.string.auto_fill_url3,
            ),
        )
    }

    override fun onDestroy() {
        super.onDestroy()
        disposable.dispose()
    }

    private fun showLoginFailed(errorString: String?) {
        Toast.makeText(applicationContext, errorString, Toast.LENGTH_SHORT).show()
    }

    companion object {
        fun getLoginActivityIntent(context: Context?): Intent {
            return Intent(context, LoginActivity::class.java)
        }
    }
}
